### PR TITLE
Add touchpad support

### PR DESCRIPTION
**Pull request by Northfear**
Touchpad support is a bit hard to get right on all devices, but once we have
- screen resolution
-  render area size and placement (it may be a window somewhere or scaled region on the screen)
-  and game resolution itself
 
most of the cases should be covered. `getWindowDestRect` and `getScreenResolution` might need overriding on Switch or some other consoles, since SDL ports may return some garbage values with them (but idea is still the same: getScreenResolution should return full resolution of the screen with touchpad and getWindowDestRect should return rect where everything is rendered (it may be a window on Linux/Windows or some centered/scaled area on console port for example).

Previously I had "trackpad" touch controls too with RMB support with 2 fingers (same as on old fheroes2 Android port), but I've dropped them later, since no one was using them anyway. I can add them too along with some option for switching between control methods, but I don't think that there's any use for it before any working mobile ports.

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @Northfear , could you please fix compilation for SDL2 on Ubuntu? It's failing for Travis-CI jobs.

File: src/engine/localevent.cpp [hunk 1]
```diff
@@ -65,6 +65,18 @@ void LocalEvent::CloseController()
         _gameController = nullptr;
     }
 }
+
+void LocalEvent::OpenTouchpad()
+{
+    int touchNumber = SDL_GetNumTouchDevices();
```
Comment (ihhub) [inline src/engine/localevent.cpp:71]: Please make this variable `const`.

File: src/engine/screen.cpp [hunk 2]
```diff
@@ -348,6 +348,25 @@ namespace
             SDL_FreeSurface( surface );
         }
 
+        virtual fheroes2::Rect getWindowDestRect() const override
+        {
+            fheroes2::Rect rect;
+            SDL_GetWindowPosition( _window, &rect.x, &rect.y );
+            SDL_GetWindowSize( _window, &rect.width, &rect.height );
+            return rect;
```
Comment (ihhub) [inline src/engine/screen.cpp:356]: We should cache this value at the time of setting resolution or changing from/to fullscreen as this operation is not instant.

File: src/engine/screen.cpp [hunk 3]
```diff
@@ -348,6 +348,25 @@ namespace
...
+        }
+
+        virtual std::pair<int, int> getScreenResolution() const override
```
Comment (ihhub) [inline src/engine/screen.cpp:359]: Please use `fheroes2::Size` type instead of std::pair.

File: src/engine/screen.cpp [hunk 4]
```diff
@@ -348,6 +348,25 @@ namespace
...
+        {
+            std::pair<int, int> resolution;
+            int32_t displayIndex = SDL_GetWindowDisplayIndex( _window );
```
Comment (ihhub) [inline src/engine/screen.cpp:362]: Please make this variable `const`.

File: src/engine/screen.cpp [hunk 5]
```diff
@@ -348,6 +348,25 @@ namespace
...
+            SDL_DisplayMode displayMode;
+            SDL_GetCurrentDisplayMode( displayIndex, &displayMode );
+            resolution.first = displayMode.w;
+            resolution.second = displayMode.h;
+            return resolution;
```
Comment (ihhub) [inline src/engine/screen.cpp:367]: We should cache this variable at the time of setting resolution or changing from/to fullscreen mode.

File: src/engine/localevent.cpp [hunk 6]
```diff
@@ -686,6 +705,58 @@ bool LocalEvent::HandleEvents( bool delay, bool allowExit )
...
+        }
+    }
+    else if ( event.type == SDL_FINGERUP ) {
+        --_numTouches;
+    }
+
+    if ( _firstFingerId == event.fingerId ) {
+        const fheroes2::Display & display = fheroes2::Display::instance();
+        const std::pair<int, int> screenResolution = fheroes2::engine().getScreenResolution(); // current resolution of screen
+        const std::pair<int, int> gameSurfaceRes( display.width(), display.height() ); // native game (surface) resolution
```
Comment (ihhub) [inline src/engine/localevent.cpp:726]: Please use `fheroes2::Size` structure instead.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Northfear , I left multiple comment here. Could you please take a look?

**Comment by Northfear** [comment]
@ihhub fixed. overriding/macroing `cacheScreenWindowRect()` should make it functional on other platforms too (switch/vita/android, etc)

File: src/engine/screen.cpp [hunk 7]
```diff
@@ -549,6 +562,18 @@ namespace
             return ( _window != NULL ) && ( ( SDL_GetWindowFlags( _window ) & SDL_WINDOW_MOUSE_FOCUS ) == SDL_WINDOW_MOUSE_FOCUS );
         }
 
+        virtual void cacheScreenWindowRect()
```
Comment (ihhub) [inline src/engine/screen.cpp:565]: This method shouldn't be virtual and should be places under private section. Please also rename it into `void _retrieveWindowInfo() const`.

File: src/engine/screen.cpp [hunk 8]
```diff
@@ -560,6 +585,8 @@ namespace
 
         std::string _previousWindowTitle;
         fheroes2::Point _prevWindowPos;
+        fheroes2::Size _screenResolution;
+        fheroes2::Rect _windowRectangle;
```
Comment (ihhub) [inline src/engine/screen.cpp:589]: Please rename it into `_activeWindowROI`.

File: src/engine/screen.cpp [hunk 9]
```diff
@@ -560,6 +585,8 @@ namespace
 
         std::string _previousWindowTitle;
         fheroes2::Point _prevWindowPos;
+        fheroes2::Size _screenResolution;
```
Comment (ihhub) [inline src/engine/screen.cpp:588]: Please rename it into `_currentScreenResolution`.

File: src/engine/screen.h [hunk 10]
```diff
@@ -52,6 +52,16 @@ namespace fheroes2
 
         virtual void setIcon( const Image & ) {}
 
+        virtual fheroes2::Rect getWindowDestRect() const
```
Comment (ihhub) [inline src/engine/screen.h:55]: Please rename it into `getActiveWindowROI`.

File: src/engine/screen.h [hunk 11]
```diff
@@ -52,6 +52,16 @@ namespace fheroes2
...
+        {
+            return fheroes2::Rect();
+        }
+
+        virtual fheroes2::Size getScreenResolution() const
```
Comment (ihhub) [inline src/engine/screen.h:60]: Please rename it into `getCurrentScreenResolution`.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Northfear , everything looks good. Please take a look into comments about names of variables and functions. I know that this is silly sometimes but we should make this core code be understandable by everyone as it's an essential part of the project.

**Comment by Northfear** [comment]
@ihhub done. `void _retrieveWindowInfo()` shouldn't be const probably, but other than I've fixed everything else.

File: src/engine/engine.cpp [hunk 12]
```diff
@@ -84,8 +84,10 @@ void SDL::Quit( void )
         Cdrom::Close();
 #endif
 #if SDL_VERSION_ATLEAST( 2, 0, 0 )
-    if ( SubSystem( SDL_INIT_GAMECONTROLLER ) )
+    if ( SubSystem( SDL_INIT_GAMECONTROLLER ) ) {
         LocalEvent::Get().CloseController();
+    }
+    LocalEvent::Get().OpenTouchpad();
```
Comment (ihhub) [inline src/engine/engine.cpp:90]: This is bug: we open a touchpad while quitting.

**Review by ihhub — CHANGES_REQUESTED** [review]

**Comment by ihhub** [comment]
Hi @Northfear I fixed a bug when we were initializing touchpad at the time of quitting the application :)

**Comment by ihhub** [comment]
@Northfear thanks a lot for such nice addition to the proejct!

**Pull request merged**